File: components/digid_x/app/models/authenticators/app_authenticator.rb

# Deze broncode is openbaar gemaakt vanwege een Woo-verzoek zodat deze
# gericht is op transparantie en niet op hergebruik. Hergebruik van 
# de broncode is toegestaan onder de EUPL licentie, met uitzondering 
# van broncode waarvoor een andere licentie is aangegeven.
# 
# Het archief waar dit bestand deel van uitmaakt is te vinden op:
#   https://github.com/MinBZK/woo-besluit-broncode-digid
# 
# Eventuele kwetsbaarheden kunnen worden gemeld bij het NCSC via:
#   https://www.ncsc.nl/contact/kwetsbaarheid-melden
# onder vermelding van "Logius, openbaar gemaakte broncode DigiD" 
# 
# Voor overige vragen over dit Woo-besluit kunt u mailen met open@logius.nl
# 
# This code has been disclosed in response to a request under the Dutch
# Open Government Act ("Wet open Overheid"). This implies that publication 
# is primarily driven by the need for transparence, not re-use.
# Re-use is permitted under the EUPL-license, with the exception 
# of source files that contain a different license.
# 
# The archive that this file originates from can be found at:
#   https://github.com/MinBZK/woo-besluit-broncode-digid
# 
# Security vulnerabilities may be responsibly disclosed via the Dutch NCSC:
#   https://www.ncsc.nl/contact/kwetsbaarheid-melden
# using the reference "Logius, publicly disclosed source code DigiD" 
# 
# Other questions regarding this Open Goverment Act decision may be
# directed via email to open@logius.nl

# frozen_string_literal: true

module Authenticators
  class AppAuthenticator < ActiveRecord::Base
    include HasBlockingManager
    include NsClient

    module Status
      ACTIVE    = "active"
      INACTIVE  = "inactive"
      PENDING   = "pending"
    end

    include Activatable

    belongs_to :account
    has_many :attempts, -> { where(attempt_type: "activation") }, as: :attemptable, dependent: :destroy

    # Internals
    validates :user_app_id, uniqueness: { case_sensitive: true }, presence: true

    # User defined
    validates :device_name, presence: true, if: -> { state.active? }

    validate :valid_pip_and_signature?, if: -> { signature_of_pip != nil }

    default_value_for :status, "inactive"
    default_value_for(:user_app_id) { SecureRandom.uuid }

    scope :most_recent, -> { order(:created_at).limit(1) }
    scope :activated_by_letter, -> { where(issuer_type: ["letter", "letter_international", "letter_secure_delivery", "letter_personal_delivery", "gemeentebalie"]) }
    scope :expired, lambda {
                      pending.where("`geldigheidstermijn` IS NULL AND DATEDIFF(?, `created_at`) > ?", Time.zone.now, ::Configuration.get_int("geldigheid_brief"))
                             .or(pending.where("DATEDIFF(?, `updated_at`) > `geldigheidstermijn`", Time.zone.now))
                    }
    scope :ordered, -> { order(activated_at: :desc) }
    scope :active_or_pending, -> { where(status: [Status::ACTIVE, Status::PENDING]) }

    scope :least_recent, -> { active.order(:last_sign_in_at).first }

    delegate :active?, to: :state
    delegate :pending?, to: :state

    def app_code
      read_attribute(:instance_id).[](0..5)&.upcase if read_attribute(:instance_id)
    end

    def device_name
      read_attribute(:device_name).truncate(35) if read_attribute(:device_name)
    end

    def self.clean_up
      delete_all.tap do
        Attempt.clean_up_missing(self)
      end
    end

    def max_number_of_failed_attempts
      ::Configuration.get_int("pogingen_activationcode_app")
    end

    def canonical_type
      self.class.name.demodulize.underscore
    end

    def activate!
      activate_with(type: :sms_controle)
    end

    def activate_with_password!
      activate_with(type: :password)
    end

    def activate_with_substantieel!
      activate_with(type: :rda)
    end

    def activate_with_activation_code!(destroy_pending = false)
      activate_with(type: :activation_code, destroy_pending: destroy_pending)
    end

    def activate_with_app!
      activate_with(type: issuer_type) # activate with current issuer_type
    end

    def mark_as_pending!(destroy_pending = false)
      account.blocking_manager_app.reset!
      transaction do
        self.status = Status::PENDING
        self.requested_at = Time.zone.now
        save!
      end

      self.class.pending.where(account_id: account_id).where.not(id: id).destroy_all if destroy_pending
    end

    def substantieel?
      !substantieel_activated_at.nil?
    end

    def hoog?
      !wid_activated_at.nil?
    end

    def valid_pip_and_signature?
      unless validate_pip_signature(signature: self.signature_of_pip)
        errors.add(:valid_pip_and_signature, "signature_of_pip is not valid")
      end
    end

    def self.substantieel?
      where("substantieel_activated_at IS NOT NULL").any?
    end

    def authentication_level
      return Account::LoginLevel::SMARTCARDPKI if hoog?
      return Account::LoginLevel::SMARTCARD if substantieel?

      Account::LoginLevel::TWO_FACTOR
    end

    def substantieel_activated_at
      self[:substantieel_activated_at] ||  self[:wid_activated_at]
    end

    def substantieel_document_type
      self[:substantieel_document_type] ||  self[:wid_document_type]
    end

    private

    def validate_pip_signature(signature:)
      false
    end

    def max_amount_of_apps
      @max_amount_of_apps ||= ::Configuration.get_int("Maximum_aantal_DigiD_apps_eindgebruiker")
    end

    def activate_with(type:, destroy_pending: false)
      account.blocking_manager_app.reset!
      transaction do
        self.status = Status::ACTIVE
        self.activated_at = Time.zone.now
        if type == :activation_code
          self.activation_code = nil
          self.geldigheidstermijn = nil
        else
          self.requested_at = Time.zone.now
          self.issuer_type = type.to_s
        end

        if account.app_authenticators.active.count >= max_amount_of_apps
          app = account.app_authenticators.least_recent
          ns_client.deregister_app(app)
          Log.instrument("1449", account_id: account_id, app_code: app.app_code, device_name: app.device_name, last_sign_in_at: I18n.l((app.last_sign_in_at || app.created_at).to_date))
          app.destroy
        end

        account.app_authenticators.pending.where.not(id: id).destroy_all if destroy_pending

        save!
      end
    end
  end
end

File: components/digid_app/src/main/java/nl/logius/digid/app/domain/authenticator/AppAuthenticator.java

/*
  Deze broncode is openbaar gemaakt vanwege een Woo-verzoek zodat deze
  gericht is op transparantie en niet op hergebruik. Hergebruik van 
  de broncode is toegestaan onder de EUPL licentie, met uitzondering 
  van broncode waarvoor een andere licentie is aangegeven.
  
  Het archief waar dit bestand deel van uitmaakt is te vinden op:
    https://github.com/MinBZK/woo-besluit-broncode-digid
  
  Eventuele kwetsbaarheden kunnen worden gemeld bij het NCSC via:
    https://www.ncsc.nl/contact/kwetsbaarheid-melden
  onder vermelding van "Logius, openbaar gemaakte broncode DigiD" 
  
  Voor overige vragen over dit Woo-besluit kunt u mailen met open@logius.nl
  
  This code has been disclosed in response to a request under the Dutch
  Open Government Act ("Wet open Overheid"). This implies that publication 
  is primarily driven by the need for transparence, not re-use.
  Re-use is permitted under the EUPL-license, with the exception 
  of source files that contain a different license.
  
  The archive that this file originates from can be found at:
    https://github.com/MinBZK/woo-besluit-broncode-digid
  
  Security vulnerabilities may be responsibly disclosed via the Dutch NCSC:
    https://www.ncsc.nl/contact/kwetsbaarheid-melden
  using the reference "Logius, publicly disclosed source code DigiD" 
  
  Other questions regarding this Open Goverment Act decision may be
  directed via email to open@logius.nl
*/

package nl.logius.digid.app.domain.authenticator;

import nl.logius.digid.app.shared.*;
import org.hibernate.annotations.DynamicUpdate;

import javax.persistence.Entity;
import javax.persistence.Table;
import java.io.Serializable;
import java.security.NoSuchAlgorithmException;
import java.time.ZonedDateTime;

@Entity
@Table(name = "app_authenticators")
@DynamicUpdate
public class AppAuthenticator extends SQLObject implements Serializable {
    private Long accountId;
    private String userAppId;
    private String status;
    private String deviceName;
    private ZonedDateTime lastSignInAt;
    private ZonedDateTime activatedAt;
    private String instanceId;
    private String userAppPublicKey;
    private String symmetricKey;
    private String maskedPin;
    private ZonedDateTime substantieelActivatedAt;
    private Boolean hardwareSupport;
    private Boolean nfcSupport;
    private String substantieelDocumentType;
    private String issuerType;
    private ZonedDateTime requestedAt;
    private String activationCode;
    private String geldigheidstermijn;
    private String pip;
    private String signatureOfPip;

    private ZonedDateTime widActivatedAt;
    private String widDocumentType;

    public AppAuthenticator() {
        this.status = "initial";
    }

    public Long getAccountId() {
        return accountId;
    }

    public void setAccountId(Long accountId) {
        this.accountId = accountId;
    }

    public String getUserAppId() {
        return userAppId;
    }

    public void setUserAppId(String userAppId) {
        this.userAppId = userAppId;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getDeviceName() {
        return sanitize(deviceName);
    }

    public void setDeviceName(String deviceName) {
        this.deviceName = sanitize(deviceName);
    }

    public ZonedDateTime getLastSignInAt() {
        return lastSignInAt;
    }

    public void setLastSignInAt(ZonedDateTime lastSignInAt) {
        this.lastSignInAt = lastSignInAt;
    }

    public ZonedDateTime getLastSignInOrActivatedAtOrCreatedAt() {
        if (this.getLastSignInAt() != null) {
            return this.getLastSignInAt();
        }
        else if (this.getActivatedAt() != null) {
            return this.getActivatedAt();
        }

        return this.getCreatedAt();
    }

    public ZonedDateTime getActivatedAt() {
        return activatedAt;
    }

    public void setActivatedAt(ZonedDateTime activatedAt) {
        this.activatedAt = activatedAt;
    }

    public void setInstanceId(String instanceId) {
        this.instanceId = instanceId;
    }

    public String getInstanceId() {
        return instanceId;
    }

    public String getUserAppPublicKey() {
        return userAppPublicKey;
    }

    public void setUserAppPublicKey(String userAppPublicKey) {
        this.userAppPublicKey = userAppPublicKey;
    }

    public String getSymmetricKey() {
        return symmetricKey;
    }

    public void setSymmetricKey(String symmetricKey) {
        this.symmetricKey = symmetricKey;
    }

    public String getMaskedPin() {
        return maskedPin;
    }

    public void setMaskedPin(String maskedPin) {
        this.maskedPin = maskedPin;
    }

    public void setSubstantieelActivatedAt(ZonedDateTime substantieelActivatedAt) {
        this.substantieelActivatedAt = substantieelActivatedAt;
    }

    public void setHardwareSupport(Boolean hardwareSupport) {
        this.hardwareSupport = hardwareSupport;
    }

    public void setNfcSupport(Boolean nfcSupport) {
        this.nfcSupport = nfcSupport;
    }

    public String getIssuerType() {
        return issuerType;
    }

    public void setIssuerType(String issuerType) {
        this.issuerType = issuerType;
    }

    public void setRequestedAt(ZonedDateTime requestedAt) {
        this.requestedAt = requestedAt;
    }

    public ZonedDateTime getRequestedAt() {
        return requestedAt;
    }

    public void setActivationCode(String activationCode) {
        this.activationCode = activationCode;
    }

    public String getActivationCode() {
        return activationCode;
    }

    public void setGeldigheidstermijn(String geldigheidstermijn) {
        this.geldigheidstermijn = geldigheidstermijn;
    }

    public String getGeldigheidstermijn() {
        return geldigheidstermijn;
    }

    public Integer getAuthenticationLevel(){
        if (widActivatedAt != null) return 30;

        return substantieelActivatedAt == null ? 20 : 25;
    }

    public String getSubstantieelDocumentType() {
        return substantieelDocumentType;
    }

    public void setSubstantieelDocumentType(String substantieelDocumentType) {
        this.substantieelDocumentType = substantieelDocumentType;
    }

    public ZonedDateTime getSubstantieelActivatedAt() {
        return substantieelActivatedAt;
    }

    public Boolean getHardwareSupport() {
        return hardwareSupport;
    }

    public Boolean getNfcSupport() {
        return nfcSupport;
    }

    public String getAppCode() {
        return getInstanceId().substring(0, 5).toUpperCase();
    }

    public String getPip() {
        return pip;
    }

    public void setPip(String pip) {
        this.pip = pip;
    }

    public String getSignatureOfPip() {
        return signatureOfPip;
    }

    public void setSignatureOfPip(String signatureOfPip) {
        this.signatureOfPip = signatureOfPip;
    }

    public boolean validatePipSignature(String signatureOfPip) {
        try {
            return (pip != null && userAppPublicKey != null &&
                ChallengeService.verifySignature(Util.toHexLower(Util.toSHA256(pip)), signatureOfPip, userAppPublicKey));
        } catch (NoSuchAlgorithmException e) {
            return false;
        }
    }

    public ZonedDateTime getWidActivatedAt() {
        return widActivatedAt;
    }

    public void setWidActivatedAt(ZonedDateTime widActivatedAt) {
        this.widActivatedAt = widActivatedAt;
    }

    public String getWidDocumentType() {
        return widDocumentType;
    }

    public void setWidDocumentType(String widDocumentType) {
        this.widDocumentType = widDocumentType;
    }

    private String sanitize(String value) {
        if (value == null) return null;
        value = value.replaceAll("[^-a-zA-Z0-9 _'\"]+", "");
        return value.substring(0, Math.min(value.length(), 36));
    }
}
